OrbitLinuxTracing/ContextSwitchAndThreadStateVisitor.cpp

```diff
@@ -0,0 +1,211 @@
+    // exiting, it has value -1. In such cases, use the association between tid and pid that we keep
+    // internally to obtain the process id.
+    pid_t prev_pid = event->GetPid();
+    if (prev_pid == -1) {
+      if (std::optional<pid_t> fallback_prev_pid = GetPidOfTid(event->GetPrevTid());
+          fallback_prev_pid.has_value()) {
+        prev_pid = fallback_prev_pid.value();
+      }
+    }
+    std::optional<SchedulingSlice> scheduling_slice = switch_manager_.ProcessContextSwitchOut(
+        prev_pid, event->GetPrevTid(), event->GetCpu(), event->GetTimestamp());
+    if (scheduling_slice.has_value()) {
+      CHECK(listener_ != nullptr);
+      listener_->OnSchedulingSlice(std::move(scheduling_slice.value()));
+    }
+  }
+
+  // Process the context switch in for scheduling slices.
+  if (event->GetNextTid() != 0) {
+    std::optional<pid_t> next_pid = GetPidOfTid(event->GetNextTid());
+    switch_manager_.ProcessContextSwitchIn(next_pid, event->GetNextTid(), event->GetCpu(),
+                                           event->GetTimestamp());
+  }
+
+  // Process the context switch out for thread state.
+  if (event->GetPrevTid() != 0 && TidMatchesPidFilter(event->GetPrevTid())) {
+    ThreadStateSlice::ThreadState new_state = GetThreadStateFromBits(event->GetPrevState());
+    std::optional<ThreadStateSlice> out_slice =
+        state_manager_.OnSchedSwitchOut(event->GetTimestamp(), event->GetPrevTid(), new_state);
+    if (out_slice.has_value()) {
+      CHECK(listener_ != nullptr);
+      listener_->OnThreadStateSlice(std::move(out_slice.value()));
+    }
+  }
+
+  // Process the context switch in for thread state.
+  if (event->GetNextTid() != 0 && TidMatchesPidFilter(event->GetNextTid())) {
+    std::optional<ThreadStateSlice> in_slice =
+        state_manager_.OnSchedSwitchIn(event->GetTimestamp(), event->GetNextTid());
+    if (in_slice.has_value()) {
+      CHECK(listener_ != nullptr);
+      listener_->OnThreadStateSlice(std::move(in_slice.value()));
+    }
+  }
+}
+
+void ContextSwitchAndThreadStateVisitor::visit(SchedWakeupPerfEvent* event) {
+  if (!TidMatchesPidFilter(event->GetWokenTid())) {
+    return;
+  }
+
+  std::optional<ThreadStateSlice> state_slice =
+      state_manager_.OnSchedWakeup(event->GetTimestamp(), event->GetWokenTid());
+  if (state_slice.has_value()) {
+    CHECK(listener_ != nullptr);
+    listener_->OnThreadStateSlice(std::move(state_slice.value()));
+  }
+}
+
+void ContextSwitchAndThreadStateVisitor::ProcessRemainingOpenStates(uint64_t timestamp_ns) {
+  std::vector<ThreadStateSlice> state_slices = state_manager_.OnCaptureFinished(timestamp_ns);
+  for (ThreadStateSlice& slice : state_slices) {
+    CHECK(listener_ != nullptr);
+    listener_->OnThreadStateSlice(slice);
+  }
+}
+
+// Associates a ThreadStateSlice::ThreadState to a thread state character retrieved from
+// /proc/<pid>/stat or the `ps` command. The possible characters are obtained from
```

Comment:
> "are obtained" here refers to a manual process, correct?

Yes yes, of course. Tweaked the comment.

> If this is done manually, should we add an error log message below when a char is encountered that we don't know about?

No error is logged because the method returns `std::nullopt` if this fails. Indeed, the one caller already logs the error.

We can also add the `ERROR` here, but that's how it's done currently.

By the way, this code wasn't touched, you see it changed because of the rename if you don't review commit by commit. That's also why you see this code but not the use of this code.